### PR TITLE
SHARP: option to set group size threshold to to use SHARP

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Mellanox CI FAILed.

**Comment by AddyLaddy** [comment]
Looks OK to me but it will leak memory when it triggers.

**Comment by AddyLaddy** [comment]
Also is SHARP useful on just 2 nodes?

**Comment by bureddy** [comment]
> Also is SHARP useful on just 2 nodes?

2 nodes it may not.  but I  guess  users just request 2 nodes and request collect,

**Comment by mellanox-github** [comment]
Mellanox CI FAILed.

File: src/sharp_plugin.c [hunk 1]
```diff
@@ -207,6 +208,24 @@ ncclResult_t ncclSharpListen(int dev, void* opaqueHandle, void** listenComm) {
 ncclResult_t ncclSharpConnect(void* handles[], int nranks, int rank, void* listenComm, void** collComm) {
   struct ncclSharpListenComm* lComm = (struct ncclSharpListenComm*)listenComm;
   struct ncclSharpCollComm* cComm;
+  char *useSharp;
+
+  if(nranks < ncclParamSharpGroupSizeThresh()) {
+    if(rank == 0) {
+      WARN("Group size:%d is less than threshold:%d. fallback to non-sharp",
```
Comment (AddyLaddy) [inline src/sharp_plugin.c:215]: Maybe we should we be consistent with using "SHARP" in the INFO/WARN messages so a simpe grep will find them

**Comment by mellanox-github** [comment]
Mellanox CI FAILed.

**Comment by mellanox-github** [comment]
Mellanox CI FAILed.

**Comment by flx42** [comment]
I think we should probably reserve `NCCL_*` variables to NCCL itself, so perhaps we should rename `NCCL_SHARP_DISABLE`? 
@AddyLaddy @sjeaugey thoughts?

**Comment by kwen2501** [comment]
Hi @flx42 , since this is a plugin for NCCL, it is okay that they prefix their environment variables with `NCCL_SHARP_*`, i.e. NCCL+plugin name. On the NCCL side, related environment variables will start with `NCCL_COLLNET_*`

**Pull request merged**